org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -219,7 +219,25 @@ public JavaProtocolExtensions getJavaExtensions() {
 	@Override
 	public void didChangeConfiguration(DidChangeConfigurationParams params) {
 		logInfo(">> workspace/didChangeConfiguration");
+		long start = System.currentTimeMillis();
+		long timeout = start + 30000;
+		boolean wait = true;
+		try {
+			while (wait) {
+				Job[] jobs = Job.getJobManager().find(InitHandler.JAVA_LS_INITIALIZATION_JOBS);
+				wait = jobs != null && jobs.length > 0 && System.currentTimeMillis() < timeout;
+			}
+		} catch (OperationCanceledException e) {
+			JavaLanguageServerPlugin.logException(e.getMessage(), e);
+		}
 		Object settings = params.getSettings();
+		JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.WORKSPACE_SYMBOL_ID, Preferences.WORKSPACE_SYMBOL);
+		JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.DOCUMENT_SYMBOL_ID, Preferences.DOCUMENT_SYMBOL);
+		JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.CODE_ACTION_ID, Preferences.CODE_ACTION);
+		JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.DEFINITION_ID, Preferences.DEFINITION);
+		JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.HOVER_ID, Preferences.HOVER);
+		JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.REFERENCES_ID, Preferences.REFERENCES);
+		JavaLanguageServerPlugin.getInstance().registerCapability(Preferences.DOCUMENT_HIGHLIGHT_ID, Preferences.DOCUMENT_HIGHLIGHT);
```

Comment:
all those capabilities should be enabled (or disabled) according to the matching getClientPreferences().isXXXDynamicRegistrationSupported(), similar to the other capabilities

Reply:
Instead of doing a shitload of 
```java
 		if (preferenceManager.getClientPreferences().isFormattingDynamicRegistrationSupported()) {
 			if (preferenceManager.getPreferences().isJavaFormatEnabled()) {
 				registerCapability(Preferences.FORMATTING_ID, Preferences.TEXT_DOCUMENT_FORMATTING);
 			} else {
 				unregisterCapability(Preferences.FORMATTING_ID, Preferences.TEXT_DOCUMENT_FORMATTING);
 			}
 		}
```
you might as well simplify it like 

```java
if (preferenceManager.getClientPreferences().isFormattingDynamicRegistrationSupported()) {
  toggleCapability(preferenceManager.getPreferences().isJavaFormatEnabled(), Preferences.FORMATTING_ID, Preferences.TEXT_DOCUMENT_FORMATTING);
}
```
with
```
private void toggleCapability(boolean enabled, String id, String capability) {
  if (enabled) {
    registerCapability(id, capability);
  } else {
     unregisterCapability(id, capability);
   }
}
```